[DataGrid] Add support for `getRowsToExport` option to print export

zreecespieces opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Fixes https://github.com/mui/mui-x/issues/9993
Fixes #9824

I created this PR in response to [an issue](https://github.com/mui/mui-x/issues/9993) I submitted regarding the Grid Print Export feature.

There was an inconsistency between the behavior and implementation of the CSV export and Print Export. When rows are selected and CSV export is chosen then only the selected rows are exported. However, when attempting to export selected rows using the Print option then all of the rows were being exported and the selection was not being honored.

This PR fixes this issue and attempts to make the implementations more similar and consistent.

- Before: https://codesandbox.io/s/cocky-booth-mw47xv
- After: https://codesandbox.io/s/falling-forest-vjlqsk

MBilalShafi reviewed (commented):
Thank you for submitting the solution, the solution looks clean and solid to me 🎉 

The only question that comes in my mind is, should we persist the checkboxes in the print version (like we did before this change), at least when there is no selection (possibly by skipping calling the `updateGridRowsForPrint` when there's no selection.) Will that be a user's expectation in some use-cases?

P.S You might need to run `yarn docs:api` to regenerate the docs and fix the build.

zreecespieces commented:
Sorry for such a late reply - I've been swamped at work and traveling so just haven't had much free time until now.

That's a good point to discuss. However, it seems that the current behavior regarding preserving checkboxes already does not include the checkboxes on print export. As of the time of my writing it seems that regardless of whether or not rows are selected the checkboxes are **NOT** preserved in the print export. I think that this behavior actually more closely matches user expectations for the following reasons.

First, the checkboxes are already not preserved when exporting as CSV and so omitting them for the print export would maintain consistency between the two methods. Second, generally users export data for reporting, presentation, or analysis. Visual elements like checkboxes don't carry data significance and thus are not necessary in the exported content. IMO the checkboxes clutter the data representation unnecessarily as the user is anticipating getting the data, not the UI elements. Third, in a digital interface a checkbox provides an interactive element for selection. In a printed or exported static format its functional relevance is lost. These all lead me to believe that omitting the checkboxes for print export would deliver the best user experience.

As I mentioned it seems that is now the current behavior, anyways. Here's a [sandbox](https://codesandbox.io/s/adoring-chandrasekhar-gd8kpc) using the latest package version (without my code changes). Notice that when exporting via print the checkboxes are never included.

I do think that only calling `updateGridRowsForPrint` if there are selected rows makes sense so I have implemented that change.

Although not the typical use case I do believe providing an option to preserve the checkboxes, such as `includeCheckboxes`, that defaults to false, would be the best solution in terms of flexibility and customization for the end user. As such I have implemented that functionality as well. Let me know what you think.

I also fixed an issue in my initial implementation where the Grid footer was not being correctly placed when exporting selected rows via print. Instead of appearing directly underneath the selected rows it appeared at the bottom of the page where it would have been had no rows been selected. This is now fixed and appearing as expected in the correct place.

I have ran the appropriate commands to ensure code formatting and new docs generation as well as syncing up with the master branch so hopefully this PR is ready for review. If there's any other issues or points of discussion I'd be happy to address them.

cherniavskii commented on packages/grid/x-data-grid/src/hooks/features/export/useGridPrintExport.tsx:
```diff
@@ -84,18 +93,37 @@ export const useGridPrintExport = (
 
+  const updateGridRowsForPrint = React.useCallback(
+    (rowsToExport?: (params: GridPrintGetRowsToExportParams<GridApiCommunity>) => GridRowId[]) =>
+      new Promise<void>((resolve) => {
```
Is there a reason for `updateGridRowsForPrint` to be asynchronous?

zreecespieces replied:
The reason I did it like that is because the `updateGridColumnsForPrint` function was setup that way and had these comments above it: `// Returns a promise because updateColumns triggers state update and
  // the new state needs to be in place before the grid can be sized correctly`

So I did it the same to be consistent - not sure if its necessary or not. I just tested removing the promise and it seems to function the same but I don't think I know enough of the context around this to understand if that would mess anything else up or not.

cherniavskii replied:
I think it's unnecessary, you can remove it 🙂

cherniavskii commented on packages/grid/x-data-grid/src/hooks/features/export/useGridPrintExport.tsx:
```diff
@@ -279,18 +321,28 @@ export const useGridPrintExport = (
+      );
+
+      const hasSelectedRows = apiRef.current.getSelectedRows().size > 0;
+      if (hasSelectedRows) {
```
Changing the default behavior would be a breaking change.
Could you keep the default behavior and add the [`getRowsToExport` option](https://mui.com/x/react-data-grid/export/#exported-rows) to `printOptions`?
This way, we'll have the same API for CSV, Excel, and Print export 🙂

zreecespieces replied:
I'm not sure what you mean, which default behavior is being changed? The `getRowsToExport` function is a method on the `GridPrintExportOptions` already, that's how I set it up initially in my first commit which is why it may not be obvious here, unless you mean something else. But [here's where that is being set](https://github.com/mui/mui-x/blob/0cd371544221552b23a04226f8efecdb5346803d/packages/grid/x-data-grid/src/models/gridExport.ts#L145) on the `printOptions`. I agree about trying to keep the API between all the export methods as similar as possible which is why I set it up this way.

zreecespieces replied:
Gotcha, I understand now what you meant by default behavior. However, I don't agree that this should be the default behavior. I think that is unintuitive and does not maintain consistency with the other export options.

Currently, if I have rows selected and export via CSV only selected rows are exported. Same with Excel. Why would the user expectation for print be any different? As a long time user and instructor of MUI I can say that the typical user surely expects the print export to behave the same and I found it confusing when that was not the case. I'm not the only one as [another issue](https://github.com/mui/mui-x/issues/9824) was also raised regarding this.

If we were to leave the current behavior and rely solely on `getRowsToExport` for this functionality the solution to achieve the same effect of only exporting selected rows would look like this:
`slotProps={{ toolbar: { printOptions: { getRowsToExport: ({ apiRef }) => Array.from(apiRef.current.getSelectedRows().keys()) } }}}`

This is extremely unintuitive and requires far greater understanding of the internal workings of the component and of JavaScript itself (knowing that `.keys()` returns a map iterator and must be converted to an array using `Array.from()`). It's also extremely verbose for such a simple feature. I do not think this is an acceptable solution for the typical user.

I think the vast majority of users expect that if they have some rows selected then only those rows will be exported regardless of the export method. I think having different behavior between the different export methods only serves to confuse users and add more complexity to what should be quite a simple functionality.

I would like to hear others' opinions on this as well.

romgrk replied:
> Gotcha, I understand now what you meant by default behavior. However, I don't agree that this should be the default behavior.

Regardless if we agree with you or not on that, we need to conform to [semver rules](https://semver.org/) to avoid breaking existing users' code (like any package on npm should do). So we're not going to include a change that breaks current code.

However, maybe there's an opportunity to refactor things a bit here, and allow passing `printOptions` from the root of the grid? This would make things more ergonomics for developpers.

```tsx
<DataGrid
  {...}
  printOptions={{ getRowsToExport: () => ... }}
/>
```

zreecespieces replied:
> Regardless if we agree with you or not on that, we need to conform to [semver rules](https://semver.org/) to avoid breaking existing users' code (like any package on npm should do). So we're not going to include a change that breaks current code.

@romgrk - Fair enough. I understand that is a valid concern. I am earnestly curious about the original intention behind this behavior and the reasoning behind its discrepancy with the other export options if this is indeed an intended feature and not a bug. But oh well.

Accepting that, I would say that I think `printOptions` is fine where it is as a prop to the `GridToolbar` as that's where `csvOptions` and `excelOptions` live as well. Since they all concern export options and the export functionality lives within the `GridToolbar` I think that is all fine and doesn't necessitate moving up to a base prop of the `DataGrid`.

My issue remains more with how verbose the solution would be for developers to implement this functionality using solely `getRowsToExport` having to do something like: `getRowsToExport: ({ apiRef }) => Array.from(apiRef.current.getSelectedRows().keys())` if they wanted to only export selected rows via print.

Given that, I propose that I refactor my current solution to preserve the current functionality (not introduce the breaking change) and instead add another option to `printOptions` called something like `exportOnlySelected` that implements this same functionality. That would be a much simpler option for developers and require far less understanding of the internal workings of the DataGrid as opposed to the `getRowsToExport` solution above.

What do you think?

bogdancbs replied:
@cherniavskii i agree with @zreecespieces, the print functionality now is not intuitive. The correct functionality would be like export csv or excel. When we select some rows at print export should be printed only selected, if nothing selected, print all. I personally need this fix asap or if you can tell me how to resolve it with a workaround it would be great.

zreecespieces replied:
@cherniavskii no worries. I think that sounds great. I will update the PR in the morning 👍🏻

zreecespieces replied:
PR has been updated based on this feedback. Current behavior is preserved, `getRowsToExport` and `includeCheckboxes` have been added to `printOptions` and documentation has been added for `printOptions` and using `getRowsToExport` to export only selected rows via print.

Let me know if you have any other feedback that needs to be addressed.

cherniavskii reviewed (approved):
Thank you!

cherniavskii commented:
I've opened https://github.com/mui/mui-x/issues/10410 to change the default behavior in v7, as we discussed 👍🏻

zreecespieces commented:
amazing. Happy to help!
